taskgroup.go

```diff
@@ -40,72 +54,58 @@ func New(ef ErrorFunc) *Group { return &Group{onError: ef} }
 
 // Go runs task in a new goroutine in g, and returns g to permit chaining.
 func (g *Group) Go(task Task) *Group {
+	if g.active.Load() == 0 {
+		g.activate()
+	}
 	g.wg.Add(1)
-	g.init()
-	errc := g.errc
 	go func() {
 		defer g.wg.Done()
 		if err := task(); err != nil {
-			errc <- err
+			g.handleError(err)
 		}
 	}()
 	return g
 }
 
-func (g *Group) init() {
-	// The first time a task is added to an otherwise clear group, set up the
-	// error collector goroutine.  We don't do this in the constructor so that
-	// an unused group can be abandoned without orphaning a goroutine.
-	g.setup.Do(func() {
-		if g.onError == nil {
-			g.onError = func(e error) error { return e }
-		}
-		g.err = nil
-		g.edone = make(chan struct{})
-		g.reset = sync.Once{}
-
-		errc := make(chan error)
-		g.errc = errc
-		go func() {
-			defer close(g.edone)
-			for err := range errc {
-				e := g.onError(err)
-				if e != nil && g.err == nil {
-					g.err = e // capture the first error always
-				}
-			}
-		}()
-	})
-}
-
-func (g *Group) cleanup() {
-	g.reset.Do(func() {
-		g.wg.Wait()
-		if g.errc == nil {
-			return
-		}
-		close(g.errc)
-		<-g.edone
-		g.errc = nil
-		g.setup = sync.Once{}
-	})
+func (g *Group) handleError(err error) {
+	g.μ.Lock()
+	defer g.μ.Unlock()
+	e := g.onError.filter(err)
+	if e != nil && g.err == nil {
+		g.err = e // capture the first unfiltered error always
+	}
 }
 
 // Wait blocks until all the goroutines currently active in the group have
-// returned, and all reported errors have been delivered to the callback.
-// It returns the first non-nil error returned by any of the goroutines in the
+// returned, and all reported errors have been delivered to the callback.  It
+// returns the first non-nil error reported by any of the goroutines in the
 // group and not filtered by an ErrorFunc.
 //
-// It is safe to call Wait concurrently from multiple goroutines, but as with
-// sync.WaitGroup no tasks can be added to g while any call to Wait is in
-// progress. Once all Wait calls have returned, the group is ready for reuse.
-func (g *Group) Wait() error { g.cleanup(); return g.err }
+// As with sync.WaitGroup, new tasks can be added to g during a call to Wait
+// only if there was already at least one task active when Wait was called.
```

Comment:
Does this need to be stronger? Any new Go must occur while at least one task is continuously active until after Go has returned.

As stated in the comment, I believe it permits the sequence:
 - G1 calls g.Go, spawns task G2
 - G1 calls g.Wait (G2 is alive, contract is satisfied)
 - G3 calls g.Go, gets through the activation check, is about to call wg.Add
 - G2 completes and calls wg.Done concurrently with the prior Add
 - wg.Add observes a zero bounce and panics

The required ordering is quite annoying to describe cleanly :/

"new tasks can be added to g during a call to Wait only if the group continuously contains at least one active task at all times, starting before the call to Wait and until after the final concurrent call to Go returns" ? :/

Reply:
You're right, I was focusing on the Wait'er too much.

Reply:
How about:
```
// As with sync.WaitGroup, new tasks can be added to g during a call to Wait
// only if the group contains at least one active task when Wait is called and
// continuously thereafter until the last concurrent call to g.Go returns.
```
?

Reply:
LGTM!